// scrapfly-sdk/script/webhookserver.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceType = void 0;
exports.verifySignature = verifySignature;
exports.createServer = createServer;
const node_buffer_1 = require("node:buffer");
const node_crypto_1 = __importDefault(require("node:crypto"));
const logger_js_1 = require("./logger.js");
var ResourceType;
(function (ResourceType) {
    ResourceType["SCRAPE"] = "scrape";
    ResourceType["PING"] = "ping";
})(ResourceType || (exports.ResourceType = ResourceType = {}));
function verifySignature(body, signature, signingSecret) {
    const secret = node_buffer_1.Buffer.from(signingSecret, 'hex');
    const hmac = node_crypto_1.default.createHmac('sha256', secret);
    hmac.update(body);
    const computed = hmac.digest('hex').toUpperCase();
    return computed === signature;
}
async function createServer(callback, app) {
    let express;
    try {
        const expressModule = await Promise.resolve().then(() => __importStar(require('express')));
        express = expressModule.default || expressModule;
    }
    catch (e) {
        throw new Error('express is not installed, please install it with `npm install express`');
    }
    if (!app) {
        app = express();
    }
    const jsonMiddleware = express.json({
        verify: (req, _res, buf) => {
            req.rawBody = buf;
        },
    });
    app.post('/webhook', jsonMiddleware, (req, res) => {
        try {
            const resourceType = req.headers['x-scrapfly-webhook-resource-type'];
            if (resourceType !== 'scrape' && resourceType !== 'ping') {
                logger_js_1.log.error(`Unsupported resource type: ${resourceType}`);
                return res.status(400).send(`this webhook server only supports scrape and ping calls`);
            }
            const data = req.body;
            try {
                callback(data, resourceType, req);
                res.status(200).send('');
            }
            catch (e) {
                logger_js_1.log.error('Callback error:', e);
                res.status(500).send('');
            }
        }
        catch (e) {
            logger_js_1.log.error('Webhook processing error:', e);
            res.status(400).send('');
        }
    });
    return app;
}
