// scrapfly-sdk/esm/webhookserver.js
import { Buffer } from 'node:buffer';
import crypto from 'node:crypto';
import { log } from './logger.js';
export var ResourceType;
(function (ResourceType) {
    ResourceType["SCRAPE"] = "scrape";
    ResourceType["PING"] = "ping";
})(ResourceType || (ResourceType = {}));
export function verifySignature(body, signature, signingSecret) {
    const secret = Buffer.from(signingSecret, 'hex');
    const hmac = crypto.createHmac('sha256', secret);
    hmac.update(body);
    const computed = hmac.digest('hex').toUpperCase();
    return computed === signature;
}
export async function createServer(callback, app) {
    let express;
    try {
        const expressModule = await import('express');
        express = expressModule.default || expressModule;
    }
    catch (e) {
        throw new Error('express is not installed, please install it with `npm install express`');
    }
    if (!app) {
        app = express();
    }
    const jsonMiddleware = express.json({
        verify: (req, _res, buf) => {
            req.rawBody = buf;
        },
    });
    app.post('/webhook', jsonMiddleware, (req, res) => {
        try {
            const resourceType = req.headers['x-scrapfly-webhook-resource-type'];
            if (resourceType !== 'scrape' && resourceType !== 'ping') {
                log.error(`Unsupported resource type: ${resourceType}`);
                return res.status(400).send(`this webhook server only supports scrape and ping calls`);
            }
            const data = req.body;
            try {
                callback(data, resourceType, req);
                res.status(200).send('');
            }
            catch (e) {
                log.error('Callback error:', e);
                res.status(500).send('');
            }
        }
        catch (e) {
            log.error('Webhook processing error:', e);
            res.status(400).send('');
        }
    });
    return app;
}
